server/config/store_config.go

```diff
@@ -118,40 +94,111 @@ func (c *StoreConfig) GetRegionMaxKeys() uint64 {
 	return uint64(c.Coprocessor.RegionMaxKeys)
 }
 
-// UpdateConfig updates the config with given config map.
-func (m *StoreConfigManager) UpdateConfig(c *StoreConfig) {
-	if c == nil || m == nil {
-		return
+// StoreConfigManager is used to manage the store config.
+type StoreConfigManager struct {
+	config *StoreConfig
+	source Source
+}
+
+// NewStoreConfigManager creates a new StoreConfigManager.
+func NewStoreConfigManager(client *http.Client) *StoreConfigManager {
+	schema := "http"
+	if ts, ok := client.Transport.(*http.Transport); ok && ts.TLSClientConfig != nil && len(ts.TLSClientConfig.Certificates) > 0 {
+		schema = "https"
+	}
+
+	return &StoreConfigManager{
+		source: newTiKVConfigSource(schema, client),
+		config: &StoreConfig{},
+	}
+}
+
+// NewTestStoreConfigManager creates a new StoreConfigManager for test.
+func NewTestStoreConfigManager(whiteList []string) *StoreConfigManager {
+	return &StoreConfigManager{
+		source: newFakeSource(whiteList),
+		config: &StoreConfig{},
 	}
-	atomic.StorePointer(&m.config, unsafe.Pointer(c))
+}
+
+// Observer is used to observe the config change.
+func (m *StoreConfigManager) Observer(address string) error {
+	cfg, err := m.source.GetConfig(address)
+	if err != nil {
+		return err
+	}
+	if cfg != nil && !reflect.DeepEqual(cfg, m.config) {
+		log.Info("pd update the store config successful", zap.String("store-config", cfg.String()))
+		*m.config = *cfg
+	}
+	return nil
 }
 
 // GetStoreConfig returns the current store configuration.
 func (m *StoreConfigManager) GetStoreConfig() *StoreConfig {
-	if m == nil || m.config == nil {
+	if m == nil {
 		return nil
 	}
-	config := atomic.LoadPointer(&m.config)
-	return (*StoreConfig)(config)
+	return m.config
+}
+
+// Source is used to get the store config.
+type Source interface {
+	GetConfig(statusAddress string) (*StoreConfig, error)
+}
+
+// TiKVConfigSource is used to get the store config from TiKV.
+type TiKVConfigSource struct {
+	schema string
+	client *http.Client
+}
+
+func newTiKVConfigSource(schema string, client *http.Client) *TiKVConfigSource {
+	return &TiKVConfigSource{
+		schema: schema,
+		client: client,
+	}
 }
 
-// Load Loads the store configuration.
-func (m *StoreConfigManager) Load(statusAddress string) error {
-	url := fmt.Sprintf("%s://%s/config", m.schema, statusAddress)
-	resp, err := m.client.Get(url)
+// GetConfig returns the store config from TiKV.
+func (s TiKVConfigSource) GetConfig(statusAddress string) (*StoreConfig, error) {
+	url := fmt.Sprintf("%s://%s/config", s.schema, statusAddress)
```

Comment:
Can we directly use the status address with schema so that we don't need to tell if it is `http/https`?

Reply:
tidb and other tools will not support?

Reply:
I guess this is only for TiKV, right?

Reply:
yes